service/frontend/adminHandler.go

```diff
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
 		return nil, adh.error(errRequestNotSet, scope)
 	}
 
-	if len(request.GetSearchAttribute()) == 0 {
+	if len(request.GetSearchAttributes()) == 0 {
 		return nil, adh.error(errSearchAttributesNotSet, scope)
 	}
 	if err := adh.validateConfigForAdvanceVisibility(); err != nil {
 		return nil, adh.error(errAdvancedVisibilityStoreIsNotConfigured, scope)
 	}
 
-	searchAttr := request.GetSearchAttribute()
-	currentValidAttr, _ := adh.DynamicConfigClient.GetMapValue(
-		dynamicconfig.ValidSearchAttributes, nil, searchattribute.GetDefaultTypeMap())
-	for k, v := range searchAttr {
-		if searchattribute.IsReservedField(k) || searchattribute.IsBuiltIn(k) {
-			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errKeyIsReservedBySystemMessage, k)), scope)
+	indexName := request.GetIndexName()
+	if indexName == "" {
+		indexName = adh.ESConfig.GetVisibilityIndex()
+	}
+
+	currentSearchAttributes, err := adh.Resource.GetSearchAttributesProvider().GetSearchAttributes(indexName, true)
+	if err != nil {
+		return nil, adh.error(serviceerror.NewInternal(fmt.Sprintf(errUnableToGetSearchAttributesMessage, err)), scope)
+	}
+
+	for saName, saType := range request.GetSearchAttributes() {
+		if searchattribute.IsReservedField(saName) {
+			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errSearchAttributeIsReservedMessage, saName)), scope)
 		}
-		if _, exist := currentValidAttr[k]; exist {
-			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errKeyIsAlreadyWhitelistedMessage, k)), scope)
+		if currentSearchAttributes.IsValid(saName) {
+			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errSearchAttributeAlreadyExistsMessage, saName)), scope)
 		}
+		if _, ok := enumspb.IndexedValueType_name[int32(saType)]; !ok {
+			return nil, adh.error(serviceerror.NewInvalidArgument(fmt.Sprintf(errUnknownSearchAttributeTypeMessage, saType)), scope)
+		}
+	}
 
-		currentValidAttr[k] = int(v)
+	// Execute workflow.
+	wfParams := addsearchattributes.WorkflowParams{
+		CustomAttributesToAdd: request.GetSearchAttributes(),
+		IndexName:             indexName,
 	}
 
-	// update dynamic config
-	err := adh.DynamicConfigClient.UpdateValue(dynamicconfig.ValidSearchAttributes, currentValidAttr)
+	run, err := adh.GetSDKClient().ExecuteWorkflow(
```

Comment:
how long will this workflow take? 
it is not a good idea to turn a API call into a workflow execution and wait for the result

maybe just start the workflow and return the pointer?

Reply:
With my local tests it takes less than 1 seconds. But I agree on big ES clusters it might be longer. I would leave it as is for now because it gives better user experience. And test it with larger cluster to check if it affects run time.